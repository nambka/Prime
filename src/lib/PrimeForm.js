import React from 'react';

// Render the input form to enter a number
class PrimeForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      inputNum: '',
      answer: ''
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {
    this.setState({
      inputNum: e.target.value
    });
  }

  handleSubmit(e) {
    // alert('Submitted data: ' + this.state.inputNum);
    e.preventDefault();
    const num = this.state.inputNum
    
    // // Naive way to check a prime
    // function isPrime(number) {
    //   if(num < 2) return false;

    //   for (let k = 2; k < num; k++){
    //     if( num % k == 0){
    //       return false;
    //     }
    //   }
    //   return true;
    // }

    function isPrime(number) {
      if (number < 2) return false;
      if (number % 2 == 0) return (number == 2);
      const sqrt = Math.sqrt(number);
      // var i;
      for (let i = 3; i <= sqrt; i += 2) {
          if (number % i == 0) return false;
      }
      return true;
    }

    function findPrimeLoop(oddNumber) {
      while (!isPrime(oddNumber)) {
        oddNumber -= 2;
      }
      return oddNumber
    }

    // Find prime < input number
    function calcPrime(number) {
      var num = parseInt(number);

      if (num < 0) {
        alert('Please input a positive integer!');
        return ''
      }

      if ((num == 0) || (num == 1) || (num == 2)) {
        return 'None'
      }

      if (num == 3) {
        return '2'
      }

      if (num > Number.MAX_SAFE_INTEGER) {
        return 'Input a number less than 2^53 - 1'
      }
      
      if (num % 2 == 0) {
        num -= 1;
        if (isPrime(num)) {
          return num
        } else {
          return findPrimeLoop(num)
        }
      } else {
        if (isPrime(num)) num -= 2;
        return findPrimeLoop(num)
      }


      // Benchmark
      // Freezing browser
      // Failsafe -1 0 1 2 2.2 10e4(99991)

      // Sieve: https://www.geeksforgeeks.org/nearest-prime-less-given-number-n/
      // https://stackoverflow.com/questions/3885817/how-do-i-check-that-a-number-is-float-or-integer/20779354#20779354

      // Number.MAX_SAFE_INTEGER: (2^53 - 1) = 9,007,199,254,740,992 / 9007199254740992

      // Number.isInteger(123) //true
      // Number.isInteger(-123) //true
      // Number.isInteger(5-2) //true
      // Number.isInteger(0) //true
      // Number.isInteger(0.5) //false
      // Number.isInteger('123') //false
      // Number.isInteger(false) //false
      // Number.isInteger(Infinity) //false
      // Number.isInteger(-Infinity) //false
      // Number.isInteger(0 / 0) //false
    }

    // // Find prime <= input number
    // function calcPrime(number) {
    //   var num = parseInt(number)

    //   if (num < 0) {
    //     alert('Please input a positive integer!');
    //     return ''
    //   }

    //   if ((num == 0) || (num == 1) || (num == 2)) {
    //     return 'None'
    //   }

    //   if (num > Number.MAX_SAFE_INTEGER) {
    //     return 'Input a number less than 2^53 - 1'
    //   }
      
    //   if (num % 2 == 0) {
    //     num -= 1;
    //     if (isPrime(num)) {
    //       return num
    //     } else {
    //       return findPrimeLoop(num)
    //     }
    //   } else {
    //     return findPrimeLoop(num)
    //   }
    // }

    this.setState({
      answer: calcPrime(num)
    })    
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit} className="container-fluid nambk-container-input text-center">
        <h2 className="nambk-header">Prime Finder</h2>
        <div className="nambk-header-desc">Highest Prime Number lower than the Input Number</div>

        <label>Input a number:
          <input type="number" value={this.state.inputNum} onChange={this.handleChange} placeholder="Enter a number"/>
        </label>
        <input type="submit" value="Go"/><br/>
        <label>Highest prime #:
          <input type="text" value={this.state.answer} readOnly />   
        </label>
      </form>
    );
  }
}
export default PrimeForm;