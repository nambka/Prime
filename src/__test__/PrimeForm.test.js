import React from 'react';
import { fireEvent, render, screen } from "@testing-library/react";
import PrimeForm from '../lib/PrimeForm';
import { shallow, mount } from 'enzyme';

describe('PrimeForm: ', () => {
    beforeEach(() => {
        // const wrapper = mount(<PrimeForm />);
    });

    afterEach(() => {
        document.getElementsByTagName('html')[0].innerHTML = '';
        jest.resetModules();
        jest.resetAllMocks();
        // wrapper.unmount();
    });
    
    it('should render the form and components', () => {
        const wrapper = shallow(<PrimeForm />);
        expect(wrapper.find('form').exists()).toBe(true);
        expect(wrapper.find('#inputNumber').length).toEqual(1);
        expect(wrapper.find('#btnFindHighestPrime').length).toEqual(1);
        expect(wrapper.find('#result').length).toEqual(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('has a title of "Prime Finder"', () => {
        const wrapper = shallow(<PrimeForm />);
        expect(wrapper.find("h2").text()).toBe("Highest Prime");
        expect(wrapper).toMatchSnapshot();
    });

    it('renders one form component', () => {
      const wrapper = shallow(<PrimeForm />);
      expect(wrapper.find('#frmFindHighestPrime')).toHaveLength(1);
      expect(wrapper).toMatchSnapshot();
    });


    // Ref: https://blog.logrocket.com/getting-started-with-enzyme-for-react-a106b58fc53b/
    // React 17.0.1 error: Cannot read property 'child' of undefined 
    // Enzyme adapter 16 not fully compatible with React 17 
    // => https://github.com/enzymejs/enzyme/issues/2462
    it('should change the state on input', () => {
        const wrapper = shallow(<PrimeForm />);
        wrapper.find('#inputNumber').simulate('change', { 
            target: { value: 100 } 
        });
        expect(wrapper.find("#inputNumber").props().value).toEqual(100);
    });


    //! NOT WORKING
    // it('should call handleSubmit method when submitting the form', done => {
    //     const state = { inputNum: '' };
    //     const props = {
    //         fetchData: jest.fn(() => state)
    //     };
    //     const wrapper = mount(<PrimeForm {...props}/>);
    //     const mockSubmit = {
    //       preventDefault: jest.fn()
    //     };
    //     wrapper.find('form').simulate('submit', mockSubmit);
    //     expect(wrapper.props().fetchData).toBeCalledWith(wrapper.state());
    //     expect(mockSubmit.preventDefault).toBeCalledTimes(1);
    //     done();
    // });

    //! NOT WORKING
    // it('should alert when input is a negative number', () => {
    //     let alertSpy = jest.spyOn(window, 'alert');
    //     const mockSubmit = jest.fn();
    //     const wrapper = shallow(<PrimeForm onSubmit={mockSubmit}/>);
    //     expect(wrapper.find('form')).toHaveLength(1);
    //     const props = {
    //         preventDefault: () => {},
    //         inputNum: -2,
    //     };
    //     wrapper.find('form').simulate('submit', props);
    //     expect(alertSpy).toBeCalledWith('Please input a positive integer!');
    //     // expect(mockSubmit).toBeCalledTimes(1);
    // });

    //! NOT WORKING
    // it('returns "None" when submit button clicked with input (0 or 1 or 2)', () => {
    //     const mockSubmit = jest.fn();
    //     const wrapper = shallow(<PrimeForm onSubmit={mockSubmit}/>);
        
    //     const props = {
    //         preventDefault: () => {},
    //         inputNum: 2,
    //     }
    //     wrapper.find('form').simulate('submit', props);
    //     expect(mockSubmit).toBeCalledWith({'inputNum': 2});

    //     // wrapper.find('[type="submit"]').simulate('submit');
    //     // wrapper.find('#findHighestPrime').simulate('submit'); 
    //     // wrapper.find('form').simulate('submit'); 

    //     // wrapper.find('form').simulate('submit', {
    //     //     preventDefault: () => {},
    //     //     target: { value: 2 }
    //     // });
    //     // // console.log(wrapper.debug())
    //     // expect(mockSubmit).toHaveBeenCalled();
    //     // expect(wrapper.find('#result').innerHTML).toBe('None');  
    // });




    // // with Promise
    // it('returns "None" when submit button clicked with input (0 or 1 or 2)', async () => {
    //     const promise = Promise.resolve();
    //     const mockCallBack = jest.fn(() => promise);
    //     const wrapper = mount(<PrimeForm onSubmit={mockCallBack}/>);
    //     // const fakeEvent = { preventDefault: () => console.log('preventDefault') };

    //     wrapper.find('#inputNumber').simulate('change', { 
    //         target: { value: '2' }
    //     });
    //     expect(wrapper.find("#inputNumber").props().value).toEqual('2');

    //     // wrapper.find('[type="submit"]').simulate('submit', fakeEvent);
    //     // wrapper.find('form').simulate('submit', fakeEvent); 
    //     wrapper.find('form').simulate('submit', {
    //         preventDefault: () => {},
    //     });
    //     // wrapper.update();

    //     expect(mockCallBack).toHaveBeenCalled();
    //     expect(wrapper.find('#result').innerHTML).toBe('None');  
    //     await act(() => promise);

    //     // setTimeout(() => {
    //     //     expect(mockCallBack).toHaveBeenCalled();
    //     //     expect(wrapper.find('#result').innerHTML).toBe('None');  
    //     // }, 1000);
    // });


})