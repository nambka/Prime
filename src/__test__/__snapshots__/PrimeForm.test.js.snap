// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimeForm with no props:  has a title of "Prime Finder" 1`] = `ShallowWrapper {}`;

exports[`PrimeForm with no props:  renders one form component 1`] = `ShallowWrapper {}`;

exports[`PrimeForm with no props:  should render the form and components 1`] = `ShallowWrapper {}`;

exports[`PrimeForm:  has a title of "Prime Finder" 1`] = `ShallowWrapper {}`;

exports[`PrimeForm:  renders one form component 1`] = `ShallowWrapper {}`;

exports[`PrimeForm:  should render the form and components 1`] = `ShallowWrapper {}`;
